File: patching/recipe.py
# SPDX-License-Identifier: GPL-3.0-only
"""Construct assembly segments without loading the source firmware."""

import hashlib
import json
from dataclasses import dataclass

INTERFACE = "classic7g-file-v1"
MAX_OUTPUT = 0xC00000


def fingerprint(data):
    return {"bytes": len(data), "sha256": hashlib.sha256(data).hexdigest()}


@dataclass(frozen=True)
class Input:
    name: str
    offset: int
    size: int


class Recipe:
    def __init__(self, inputs):
        self.inputs = inputs
        self.segments = []
        self.checks = []
        self.relocations = {}
        self.ffs_checksums = []
        self.data = bytearray()
        self.size = 0

    def overlay(self, original, replacements):
        """Emit a source region or compiled image with nonoverlapping replacements."""
        size = original.size if isinstance(original, Input) else len(original)
        position = 0
        for offset, replacement in sorted(replacements, key=lambda span: span[0]):
            length = (
                replacement.size if isinstance(replacement, Input) else len(replacement)
            )
            if offset < position or length <= 0 or offset + length > size:
                raise ValueError(
                    f"Overlapping or out-of-range replacement at {offset:#x}"
                )
            self._region(original, position, offset)
            self._region(replacement, 0, length)
            position = offset + length
        self._region(original, position, size)

    def _region(self, part, start, end):
        if start == end:
            return
        if isinstance(part, Input):
            self.source(part.name, part.offset + start, end - start)
        else:
            self.literal(part[start:end])

    def relocate(self, name, base, count):
        """Source copies use relocated PE bytes; preimage checks use originals."""
        if name not in self.inputs or name in self.relocations:
            raise ValueError(f"Invalid or duplicate relocation input: {name}")
        if base < 0 or base + self.inputs[name]["bytes"] > 0x100000000 or count <= 0:
            raise ValueError(f"Invalid relocation layout: {name}")
        self.relocations[name] = {"base": base, "count": count}

    def _append(self, segment):
        size = segment["bytes"]
        if size <= 0 or self.size + size > MAX_OUTPUT:
            raise ValueError("Invalid recipe output size")
        if self.segments:
            previous = self.segments[-1]
            if (
                previous["kind"] == segment["kind"]
                and previous.get("name") == segment.get("name")
                and (
                    segment["kind"] == "zero"
                    or previous["offset"] + previous["bytes"] == segment["offset"]
                )
            ):
                previous["bytes"] += size
                self.size += size
                return
        self.segments.append(segment)
        self.size += size

    def source(self, name, offset, size):
        if (
            name not in self.inputs
            or offset < 0
            or size <= 0
            or offset + size > self.inputs[name]["bytes"]
        ):
            raise ValueError(f"Copy source outside input: {name} at {offset:#x}")
        self._append({"kind": "input", "name": name, "offset": offset, "bytes": size})

    def literal(self, data):
        # Emit long padding runs without storing them in the compiled-data asset.
        position = 0
        while position < len(data):
            zero = data.find(bytes(32), position)
            end = len(data) if zero < 0 else zero
            literal = data[position:end]
            if literal:
                self._append(
                    {"kind": "data", "offset": len(self.data), "bytes": len(literal)}
                )
                self.data.extend(literal)
            if zero < 0:
                break
            end = zero + 32
            while end < len(data) and data[end] == 0:
                end += 1
            self.zero(end - zero)
            position = end

    def zero(self, size):
        if size:
            self._append({"kind": "zero", "bytes": size})

    def expect(self, name, offset, data):
        if (
            name not in self.inputs
            or offset < 0
            or not data
            or offset + len(data) > self.inputs[name]["bytes"]
        ):
            raise ValueError("Expected bytes outside input")
        self.checks.append({"name": name, "offset": offset, "hex": data.hex()})

    def save(self, directory, name):
        if not self.segments:
            raise ValueError("Empty recipe")
        spec = {
            "schema": 2,
            "interface": INTERFACE,
            "inputs": self.inputs,
            "output": {"bytes": self.size},
            "data": fingerprint(self.data),
            "checks": self.checks,
            "relocations": self.relocations,
            "ffs_checksums": self.ffs_checksums,
            "segments": self.segments,
        }
        raw = json.dumps(spec, sort_keys=True, separators=(",", ":")) + "\n"
        if len(raw.encode()) > 2 * 1024 * 1024:
            raise ValueError("Recipe exceeds 2 MiB")
        directory.mkdir(parents=True, exist_ok=True)
        path = directory / (name + ".json")
        data_path = directory / (name + ".data")
        path.write_text(raw)
        data_path.write_bytes(self.data)
        return {"recipe": str(path), "data": str(data_path)}

File: host/crates/reprise-bundle/src/assembly/recipe.rs
// SPDX-License-Identifier: GPL-3.0-only
//! Bounded recipe validation and assembly shared by local builds and bundles.

use super::{pe, INPUT_FILES, INTERFACE};
use crate::{invalid, sha256, valid_hash, Result};
use serde::{Deserialize, Serialize};
use std::collections::BTreeMap;

pub(super) const MAX_RECIPE: usize = 2 * 1024 * 1024;
// OSOS staging space before the companion in classic7g-file-v1.
pub(super) const MAX_OUTPUT: usize = 0x03c00000 - 0x03000000;

#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(deny_unknown_fields)]
pub(super) struct Fingerprint {
    pub(super) bytes: usize,
    pub(super) sha256: String,
}

impl Fingerprint {
    pub(super) fn check(&self, bytes: &[u8]) -> Result<()> {
        if bytes.len() != self.bytes || sha256(bytes) != self.sha256 {
            return Err(invalid("Firmware input/output fingerprint mismatch"));
        }
        Ok(())
    }
}

#[derive(Debug, Deserialize)]
#[serde(tag = "kind", rename_all = "snake_case", deny_unknown_fields)]
enum Segment {
    Input {
        name: String,
        offset: usize,
        bytes: usize,
    },
    Data {
        offset: usize,
        bytes: usize,
    },
    Zero {
        bytes: usize,
    },
}

#[derive(Debug, Deserialize)]
#[serde(deny_unknown_fields)]
pub(super) struct Output {
    pub(super) bytes: usize,
}

#[derive(Debug, Deserialize)]
#[serde(deny_unknown_fields)]
struct Check {
    name: String,
    offset: usize,
    hex: String,
}

#[derive(Debug, Deserialize)]
#[serde(deny_unknown_fields)]
struct Relocation {
    base: u32,
    count: usize,
}

#[derive(Debug, Deserialize)]
#[serde(deny_unknown_fields)]
pub(super) struct Recipe {
    schema: u32,
    interface: String,
    pub(super) inputs: BTreeMap<String, Fingerprint>,
    pub(super) output: Output,
    data: Fingerprint,
    #[serde(default)]
    checks: Vec<Check>,
    #[serde(default)]
    relocations: BTreeMap<String, Relocation>,
    #[serde(default)]
    ffs_checksums: Vec<usize>,
    segments: Vec<Segment>,
}

pub(super) fn slice(data: &[u8], offset: usize, bytes: usize) -> Result<&[u8]> {
    let end = offset
        .checked_add(bytes)
        .ok_or_else(|| invalid("Recipe range overflow"))?;
    data.get(offset..end)
        .ok_or_else(|| invalid("Recipe range outside input"))
}

impl Recipe {
    pub(super) fn parse(bytes: &[u8]) -> Result<Self> {
        if bytes.len() > MAX_RECIPE {
            return Err(invalid("Assembly recipe exceeds 2 MiB"));
        }
        let recipe: Self = serde_json::from_slice(bytes)?;
        if recipe.schema != 2
            || recipe.interface != INTERFACE
            || recipe.inputs.is_empty()
            || recipe.segments.is_empty()
            || recipe.output.bytes == 0
            || recipe.output.bytes > MAX_OUTPUT
        {
            return Err(invalid("Unsupported assembly recipe"));
        }
        if recipe.data.bytes > MAX_OUTPUT || !valid_hash(&recipe.data.sha256) {
            return Err(invalid("Invalid recipe data fingerprint"));
        }
        for (name, spec) in &recipe.inputs {
            if !INPUT_FILES.iter().any(|(key, _)| key == name)
                || spec.bytes == 0
                || spec.bytes > MAX_OUTPUT
                || !valid_hash(&spec.sha256)
            {
                return Err(invalid("Unsupported Apple input"));
            }
        }
        Ok(recipe)
    }

    pub(super) fn apply(&self, inputs: &BTreeMap<String, Vec<u8>>, data: &[u8]) -> Result<Vec<u8>> {
        for (name, spec) in &self.inputs {
            spec.check(
                inputs
                    .get(name)
                    .ok_or_else(|| invalid(format!("Missing Apple input: {name}")))?,
            )?;
        }
        self.data.check(data)?;
        for check in &self.checks {
            if !self.inputs.contains_key(&check.name)
                || check.hex.is_empty()
                || check.hex.len() % 2 != 0
                || !check.hex.bytes().all(|byte| byte.is_ascii_hexdigit())
            {
                return Err(invalid("Invalid recipe preimage check"));
            }
            let expected: Vec<u8> = check
                .hex
                .as_bytes()
                .chunks_exact(2)
                .map(|pair| u8::from_str_radix(std::str::from_utf8(pair).unwrap(), 16).unwrap())
                .collect();
            if slice(&inputs[&check.name], check.offset, expected.len())? != expected {
                return Err(invalid(format!(
                    "Recipe preimage mismatch: {} at {:#x}",
                    check.name, check.offset
                )));
            }
        }
        let mut relocated = BTreeMap::new();
        for (name, relocation) in &self.relocations {
            if !self.inputs.contains_key(name) {
                return Err(invalid("Relocation uses an unverified input"));
            }
            relocated.insert(
                name,
                pe::relocate(&inputs[name], relocation.base, relocation.count)?,
            );
        }
        let mut output = Vec::with_capacity(self.output.bytes);
        for segment in &self.segments {
            let bytes = match segment {
                Segment::Input { bytes, .. }
                | Segment::Data { bytes, .. }
                | Segment::Zero { bytes } => *bytes,
            };
            if bytes == 0 || bytes > self.output.bytes - output.len() {
                return Err(invalid("Recipe exceeds output bounds"));
            }
            match segment {
                Segment::Input { name, offset, .. } => {
                    if !self.inputs.contains_key(name) {
                        return Err(invalid("Recipe uses an unverified input"));
                    }
                    let source = relocated.get(name).unwrap_or(&inputs[name]);
                    output.extend_from_slice(slice(source, *offset, bytes)?);
                }
                Segment::Data { offset, .. } => {
                    output.extend_from_slice(slice(data, *offset, bytes)?)
                }
                Segment::Zero { .. } => output.resize(output.len() + bytes, 0),
            }
        }
        if output.len() != self.output.bytes {
            return Err(invalid("Recipe output size mismatch"));
        }
        for offset in &self.ffs_checksums {
            update_ffs_checksum(&mut output, *offset)?;
        }
        Ok(output)
    }
}

pub(super) fn update_ffs_checksum(image: &mut [u8], offset: usize) -> Result<()> {
    let header = slice(image, offset, 24)?;
    let size =
        usize::from(header[20]) | usize::from(header[21]) << 8 | usize::from(header[22]) << 16;
    let header_sum = header
        .iter()
        .enumerate()
        .filter(|(i, _)| *i != 17 && *i != 23)
        .fold(0u8, |sum, (_, byte)| sum.wrapping_add(*byte));
    if size < 24 || header[18..20] != [4, 0x40] || header_sum != 0 {
        return Err(invalid("Invalid PreEfi FFS header"));
    }
    let body = &slice(image, offset, size)?[24..];
    image[offset + 17] = body.iter().fold(0u8, |sum, byte| sum.wrapping_sub(*byte));
    Ok(())
}
